registry/client/endpoint.go

```diff
@@ -0,0 +1,268 @@
+package client
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/distribution/registry/api/v2"
+)
+
+// Authorizer is used to apply Authorization to an HTTP request
+type Authorizer interface {
+	// Authorizer updates an HTTP request with the needed authorization
+	Authorize(req *http.Request) error
```

Comment:
The current role of the `Endpoint` object is too large and breaking it apart is in progress.  I don't fully agree though that we are that far from the goauth2 model and instead using the model from the engine.

Some modifications I am making
- Single `Transport` type which takes in an Authorizer and Base set of Headers (keeps us from needing a request factory)
- An `Authorizer` implementation that is able to discover the appropriate auth mechanism and pull in needed credentials and get tokens (this is your session concept)
- A configuration which takes in a base set of headers and auth methods and is used to construct an `*http.Client`